util/run.bzl

```diff
@@ -0,0 +1,199 @@
+            ctx.file.image_tar] + 
+            ctx.attr.image_tar.files.to_list() + 
+            ctx.attr.image_tar.data_runfiles.files.to_list()
+        ),
+    )
+
+def _commit_impl(ctx):
+    # Since we're always bundling/renaming the image in the macro, this is valid.
+    load_statement = 'docker load -i %s' % ctx.file.image_tar.short_path
+
+    # Generate a shell script to execute the run statement
+    ctx.actions.expand_template(
+        template=ctx.file._run_tpl,
+        output=ctx.outputs.executable,
+        substitutions={
+          "%{load_statement}": load_statement,
+          "%{flags}": " ".join(ctx.attr.flags),
+          "%{image}": ctx.attr.image_name,
+          "%{original_image}": ctx.attr.original_image,
+          "%{command}": _process_command(ctx.attr.command),
+        },
+        is_executable=True,
+    )
+
+    return struct(runfiles=ctx.runfiles(files = [
+            ctx.executable.image_tar,
+            ctx.file.image_tar] + 
+            ctx.attr.image_tar.files.to_list() + 
+            ctx.attr.image_tar.data_runfiles.files.to_list()
+        ),
+    )
+
+_run_and_commit = rule(
+    attrs = {
+        "flags": attr.string_list(
+            doc = "list of flags to pass to run command",
+            default = [],
+        ),
+        "image_tar": attr.label(
+            executable = True,
+            allow_files = True,
+            mandatory = True,
+            single_file = True,
+            cfg = "target",
+        ),
+        "original_image": attr.string(
+            doc = "name of original image (for computing sha)",
+            mandatory = True,
+        ),
+        "image_name": attr.string(
+            doc = "name of image to run commands on",
+            mandatory = True,
+        ),
+        "command": attr.string_list(
+            doc = "command to run",
+            mandatory = True,
+            non_empty = True,
+        ),
+        "_run_tpl": attr.label(
+            default = Label("//util:commit.sh.tpl"),
+            allow_files = True,
+            single_file = True,
+        ),
+    },
+    executable = True,
+    implementation = _commit_impl,
+)
+
+_run_and_extract = rule(
+    attrs = {
+        "flags": attr.string_list(
+            doc = "list of flags to pass to run command",
+            default = [],
+        ),
+        "image_tar": attr.label(
+            executable = True,
+            allow_files = True,
+            mandatory = True,
+            single_file = True,
+            cfg = "target",
+        ),
+        "image_name": attr.string(
+            doc = "name of image to run commands on",
+            mandatory = True,
+        ),
+        "command": attr.string_list(
+            doc = "command to run",
+            mandatory = True,
+            non_empty = True,
+        ),
+        "extract_file": attr.string(
+            doc = "path to file to extract from container",
+            mandatory = True,
+        ),
+        "target": attr.string(
```

Comment:
hmm, now that i think more about this, we should skip this attribute. By default this will be placed in the runfiles directory of the target.
e.g.:
bazel-bin/package_managers/apt_get/download_pkgs.runfiles/base_images_docker/package_managers/apt_get/download_pkgs.tar

The reason to avoid this are:
1. We need to add checks if the destination path is valid and deal with that.
2. Possiblity that even if the path is valid, might not be available in Bazel Sandbox env.
https://blog.bazel.build/2015/09/11/sandboxing.html

Users can copy this artifact where they want after the execution. Other rules which consume this file can also predict the path from target package namespace.

Reply:
ah yeah you're probably right, I don't really want to deal with validation. I'll remove this and just default to `.` which will put it in the runfiles directory.